mmseg/models/backbones/vit.py

```diff
@@ -0,0 +1,356 @@
+                 depth=12,
+                 num_heads=12,
+                 mlp_ratio=4,
+                 qkv_bias=True,
+                 qk_scale=None,
+                 representation_size=None,
+                 drop_rate=0.,
+                 attn_drop_rate=0.,
+                 drop_path_rate=0.,
+                 norm_cfg=dict(type='LN'),
+                 act_cfg=dict(type='GELU'),
+                 norm_eval=False,
+                 with_cp=False,
+                 weight_init=''):
+        super(VisionTransformer, self).__init__()
+        self.img_size = img_size
+        self.patch_size = patch_size
+        self.pretrain_img_size = pretrain_img_size
+        self.features = self.embed_dim = embed_dim
+        self.patch_embed = PatchEmbed(
+            img_size=img_size,
+            patch_size=patch_size,
+            in_channels=in_channels,
+            embed_dim=embed_dim)
+        num_patches = (pretrain_img_size // patch_size)**2
+
+        self.pos_embed = nn.Parameter(torch.zeros(1, num_patches, embed_dim))
+        self.pos_drop = nn.Dropout(p=drop_rate)
+
+        dpr = [x.item() for x in torch.linspace(0, drop_path_rate, depth)]
+        self.blocks = nn.Sequential(*[
+            Block(
+                dim=embed_dim,
+                num_heads=num_heads,
+                mlp_ratio=mlp_ratio,
+                qkv_bias=qkv_bias,
+                qk_scale=qk_scale,
+                drop=drop_rate,
+                attn_drop=attn_drop_rate,
+                drop_path=dpr[i],
+                act_cfg=act_cfg,
+                norm_cfg=norm_cfg) for i in range(depth)
+        ])
+        _, self.norm = build_norm_layer(norm_cfg, embed_dim)
+
+        if representation_size:
+            self.num_features = representation_size
+            self.pre_logits = nn.Sequential(
+                OrderedDict([('fc', Linear(embed_dim, representation_size)),
+                             ('act',
+                              build_activation_layer(act_cfg=dict('Tanh')))]))
+        else:
+            self.pre_logits = nn.Identity()
+
+        self.norm_eval = norm_eval
+        self.with_cp = with_cp
+        # weight init mode
+        self.weight_init = weight_init
+
+    def init_weights(self, pretrained=None):
+        if isinstance(pretrained, str):
+            logger = get_root_logger()
+            state_dict = load_checkpoint(
+                self, pretrained, strict=False, logger=logger)
+            if 'pos_embed' in state_dict.keys(
+            ) and state_dict['pos_embed'].shape != self.pos_embed.shape:
+                self.pos_embed = nn.Parameter(
+                    torch.zeros(state_dict['pos_embed'].shape))
+                logger.info(msg='Reload checkpoint')
+                load_checkpoint(self, pretrained, strict=False, logger=logger)
+                self.pos_embed = nn.Parameter(self.pos_embed[:, 1:, :])
+
+                if self.patch_embed.num_patches != self.pos_embed.shape[1]:
```

Comment:
We'd better compare `self.pos_embed.shape` against `state_dict['pos_embed'].shape`.

Reply:
In this if block, I want to resize the pretrained pos_embed weights to adapt training image, so it should compare `[B, L_training, C]` and` [B, L_pretraining, C]`. In this code, `self.patch_embed.num_patches` refers to `L_trainnig`, and `self.pos_embed.shape[1] `refers to `L_pretraining`.

Reply:
`self.patch_embed.num_patches` is equal to `self.pos_embed.shape[1]`.